File: 2019/quals/reversing-android/checker.c
// Copyright 2019 Google LLC
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//   https://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// gcc -o checker checker.c; ./checker

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int decisions[] = {0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0,
                   0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0,
                   0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0};
int place = 0;
int correct = 0;

void MergeSort(char* arr, int len) {
  if (len <= 1) {
    return;
  }
  int len1 = len / 2;
  int len2 = len - len1;
  char* l = arr;
  char* r = arr + len1;

  MergeSort(arr, len1);
  if (!correct) {
    return;
  }
  MergeSort(r, len2);
  if (!correct) {
    return;
  }

  int i = 0, j = 0, k = 0;
  char temp[16];

  while (i < len1 && j < len2) {
    if (l[i] < r[j]) {
      if (decisions[place] != 1) {
        correct = 0;
        return;
      }
      ++place;

      temp[k] = l[i];
      ++i;
    } else if (l[i] > r[j]) {
      if (decisions[place] != 0) {
        correct = 0;
        return;
      }
      ++place;

      temp[k] = r[j];
      ++j;
    } else {
      correct = 0;
      return;
    }
    ++k;
  }
  while (i < len1) {
    temp[k] = l[i];
    ++i;
    ++k;
  }
  while (j < len2) {
    temp[k] = r[j];
    ++j;
    ++k;
  }
  memcpy(arr, temp, len);
}

int Checker(char* key) {
  char sum[16];
  for (int i = 0; i < 16; ++i) {
    sum[i] = key[2 * i] + key[2 * i + 1];
  }

  correct = 1;
  place = 0;
  MergeSort(sum, 16);
  return correct;
}

int main(int argc, char **argv) {
  // Sanity check 0: Returns true on correct key
  char correct_key[] = {9, 0, 0, 8, 0,  7, 2, 0, 0, 11, 0, 15, 13, 0, 10, 0,
                        6, 0, 0, 5, 14, 0, 0, 4, 0, 3,  0, 0,  12, 0, 1,  0};
  printf("Result for right key: %d\n", Checker(correct_key));

  // Sanity check 1: We shouldn't return true for a modified key
  printf("\nSanity check 1\n");
  for (int i = 0; i < 16; ++i) {
    for (int new_char = 1; new_char < 16; ++new_char) {
      // Replace the non-0 char
      int replace = 2 * i;
      if (correct_key[replace] == 0) {
        replace = 2 * i + 1;
      }
      printf("Replacing correct_key[%d] (%d) with %d\n", replace,
             correct_key[replace], new_char);

      // Insert new_char to key[replace]
      if (new_char == correct_key[replace]) {
        continue;
      }

      char bad_key[32];
      memcpy(bad_key, correct_key, 32);

      // Elements need to be unique, so replace the other new_char with current
      // key[replace]
      for (int j = 0; j < 32; ++j) {
        if (bad_key[j] == new_char) {
          bad_key[j] = bad_key[replace];
          break;
        }
      }

      bad_key[replace] = new_char;

      printf("Checking ");
      for (int j = 0; j < 32; ++j) {
        printf("%d ", bad_key[j]);
      }
      if (Checker(bad_key)) {
        puts(": Checker returns true :C");
        return 1;
      } else {
        puts(": Checker returns false");
      }
    }
  }

  // Sanity check 2: We should still return true if the sums stay the same
  printf("\nSanity check 2\n");
  for (int replace = 0; replace < 32; replace += 2) {
    char bad_key[32];
    memcpy(bad_key, correct_key, 32);
    for (int new_char = 0; new_char < 16; ++new_char) {
      if (new_char == correct_key[replace]) {
        continue;
      }
      int correct_sum = correct_key[replace] + correct_key[replace + 1];
      int other_char = correct_sum - new_char;
      if (other_char < 0 || other_char >= 16) {
        continue;
      }
      bad_key[replace] = new_char;
      bad_key[replace + 1] = other_char;
      printf("Checking ");
      for (int j = 0; j < 32; ++j) {
        printf("%d ", bad_key[j]);
      }
      if (!Checker(bad_key)) {
        puts(": Checker returns false :C");
        return 1;
      } else {
        puts(": Checker returns true");
      }
    }
  }
  return 0;
}

File: other/re-risky/split-version/task2/spoilers_and_source/src/checker.py
#!/usr/bin/python
# Copyright 2018 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http:#www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import hashlib

def check(s):
  def get():
    def get():
      def get():
        def get():
          def get():
            def get():
              def get():
                return ["5ef3f2aae145166cbc0b738afab2f8d90c299edee5eb4051107395e3ba637a86","3fdc49f4aefc086588cf75289178a44351a85033e6c7e9e4a3337e716f3f221f","5d5b77b0dd736777b522fc19572e5a71d4bb29cfdf78979efca3dca8361bda74","7eeecf1e3339198b2a08d7738ee59ded2449be5f34057383b0f830bd19fb7968","5093702094d4c2f8b72c3b32adfea3cc150e66fd1fced73f5b6a10b8910db337","ba2853f7fbe062df0e86cd12719e2af0eda18a4583f72604a318a99a970d3fc7","29a057c4102f371edce5b2baf29f330a645092b7918a91ba1e45baecf94cda13","7eeecf1e3339198b2a08d7738ee59ded2449be5f34057383b0f830bd19fb7968","a734ff7b5532da3189d49dcf395e85244f4aaf9052ad24ec14563301c030a095","8aba9e709af71b3d2d38b50e0debe42253e076f57adfa47a4a33092ff17eb27d","fc52dad433d9cf4ca2be74096d3c9ea3c5ac7e81edc2cb44d06aeef09b52b551","2e874850ba1dc08874ddde324ae2d0701497c26ff3a7ec425b9a1a139bf717e8","3a7b0788327b208bb25af82aa02d0106f74f21b3b9d28da591422f71b460569c","4e234ce843218d3c569a5a3d1565a578ec81573401093198d4cdb62e52f00f35","ba2853f7fbe062df0e86cd12719e2af0eda18a4583f72604a318a99a970d3fc7","fc52dad433d9cf4ca2be74096d3c9ea3c5ac7e81edc2cb44d06aeef09b52b551","68b85128461450d83b626fede97057d87ce1789d80ac8012521059567d2db799",]
              return get()
            return get()
          return get()
        return get()
      return get()
    return get()

  v = get()
  if len(v) != len(s):
    return False

  return all([
    hashlib.sha256("SUFFIX" + ch + "PREFIX").hexdigest() == tst
    for ch, tst in zip(s, v)
    ])

